frontend/src/metabase/lib/redux/typed-utils.ts

```diff
@@ -11,13 +12,26 @@ interface ThunkConfig {
 export const createAsyncThunk =
   createAsyncThunkOriginal.withTypes<ThunkConfig>();
 
+type AwaitedIfPromise<T> = T extends Promise<infer R> ? R : T;
+
 // similar to createAction but accepts a (redux-thunk style) thunk and dispatches based on whether
 // the promise returned from the thunk resolves or rejects, similar to redux-promise
-export function createThunkAction<TArgs extends any[]>(
-  actionType: string,
+export function createThunkAction<
+  TArgs extends any[],
+  TResult,
+  TActionType extends string,
+>(
+  actionType: TActionType,
   thunkCreator: (
     ...args: TArgs
-  ) => (dispatch: Dispatch, getState: GetState) => any,
-): (...args: TArgs) => any {
+  ) => (dispatch: ThunkDispatch<any, any, any>, getState: GetState) => TResult,
+): (
+  ...args: TArgs
+) => (
+  dispatch: ThunkDispatch<any, any, any>,
+  getState: GetState,
+) => Promise<{ type: TActionType; payload: AwaitedIfPromise<TResult> }> {
```

Comment:
Note:
`await dispatch(thunk created in this way)`
and
`await dispatch(thunk created by createAsyncThunk from rtk)`
will return different things, the former will return what's typed there (so the resulting action), while the latter will return what's being returned by the thunk (so the payload of our function).